### PR TITLE
Improves coverage for the aws_cryptosdk_md_{init, finish} proofs 

### PR DESCRIPTION
**Pull request by tegansb**
Description of changes:
Addresses the low coverage of the aws_cryptosdk_md_init and aws_cryptosdk_md_finish proofs.
This PR builds on PR # 574. 
Includes a README for both the aws_cryptosdk_md_init and aws_cryptosdk_md_finish proofs specifying expected coverage behavior.
Uses stubs for EVP_MD_CTX_free,  evp_md_ctx_is_valid and aws_raise_error_private to improve coverage. 
Uses AWS_FATAL_PRECONDITION instead of abort().
Coverage for aws_cryptosdk_md_init is improved from .73 to 1.0.
Coverage for aws_cryptosdk_md_finish is improved from .71 to .99.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

@danielsn @feliperodri

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_cryptosdk_md_abort/Makefile [hunk 1]
```diff
@@ -35,6 +36,11 @@ DEPENDENCIES += $(SRCDIR)/helper-src/openssl/bn_override.goto
 DEPENDENCIES += $(SRCDIR)/helper-src/openssl/ec_override.goto
 DEPENDENCIES += $(SRCDIR)/helper-src/openssl/evp_override.goto
 
+REMOVE_FUNCTION_BODIES += --remove-function-body EVP_MD_CTX_free
+REMOVE_FUNCTION_BODIES += --remove-function-body evp_md_ctx_is_valid
```
Comment (tawdry-audrey) [inline .cbmc-batch/jobs/aws_cryptosdk_md_abort/Makefile:40]: If this is due to runtime/memory issues with CBMC, just add a comment explaining why these have to be abstracted to the proofs. Otherwise, looks good to me!

Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_md_abort/Makefile]: Thanks! I've updated this as per https://github.com/aws/aws-encryption-sdk-c/pull/582#discussion_r478660427 and added some documentation to the Makefiles about why we use the abstractions.

File: .cbmc-batch/stubs/EVP_MD_CTX_free_no_pkey_stub.c [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+#include <ec_utils.h>
+#include <make_common_data_structures.h>
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+#include <openssl/kdf.h>
+#include <openssl/rsa.h>
+#include <proof_helpers/nondet.h>
+
+/* Abstraction of the EVP_MD_CTX struct */
```
Comment (danielsn) [inline .cbmc-batch/stubs/EVP_MD_CTX_free_no_pkey_stub.c:24]: Why is this here, instead of in a .h file? Are we sure we'll never need it outside this file?

File: .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+#include <openssl/kdf.h>
+#include <openssl/rsa.h>
+#include <proof_helpers/nondet.h>
+
+/* Abstraction of the EVP_MD_CTX struct */
+struct evp_md_ctx_st {
+    bool is_initialized;
+    EVP_PKEY *pkey;
+    size_t digest_size;
+};
```
Comment (danielsn) [inline .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c:29]: Same as above: why is this here and not in a .h?

File: .cbmc-batch/stubs/EVP_MD_CTX_free_no_pkey_stub.c (hunk 2 above)
Comment (tegansb) [reply .cbmc-batch/stubs/EVP_MD_CTX_free_no_pkey_stub.c]: Fixed both here and in https://github.com/aws/aws-encryption-sdk-c/pull/574

File: .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c (hunk 3 above)
Comment (tegansb) [reply .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c]: Same as above

File: .cbmc-batch/include/openssl/evp.h [hunk 4]
```diff
@@ -17,14 +17,20 @@
...
-
 #include <openssl/ec.h>
 #include <openssl/ossl_typ.h>
+#include <stdbool.h>
+#include <stddef.h>
 
 #define EVP_MAX_MD_SIZE 64  /* longest known is SHA512 */
 #define EVP_PKEY_HKDF 1036  // reference from obj_mac.h
 
+/* Abstraction of the EVP_PKEY struct */
```
Comment (feliperodri) [inline .cbmc-batch/include/openssl/evp.h:28]:
```c
/* Abstraction of the EVP_PKEY struct. */
```

File: .cbmc-batch/include/openssl/evp.h [hunk 5]
```diff
@@ -45,6 +51,13 @@ int EVP_PKEY_CTX_set_rsa_mgf1_md(EVP_PKEY_CTX *ctx, const EVP_MD *md);
 int EVP_PKEY_encrypt(EVP_PKEY_CTX *ctx, unsigned char *out, size_t *outlen, const unsigned char *in, size_t inlen);
 int EVP_PKEY_decrypt(EVP_PKEY_CTX *ctx, unsigned char *out, size_t *outlen, const unsigned char *in, size_t inlen);
 
+/* Abstraction of the EVP_MD_CTX struct */
```
Comment (feliperodri) [inline .cbmc-batch/include/openssl/evp.h:54]:
```c
/* Abstraction of the EVP_MD_CTX struct. */
```

File: .cbmc-batch/include/openssl/evp.h [hunk 6]
```diff
@@ -45,6 +51,13 @@ int EVP_PKEY_CTX_set_rsa_mgf1_md(EVP_PKEY_CTX *ctx, const EVP_MD *md);
...
+struct evp_md_ctx_st {
+    bool is_initialized;
+    EVP_PKEY *pkey;
+    size_t digest_size;
+};
```
Comment (feliperodri) [inline .cbmc-batch/include/openssl/evp.h:59]: Could you group struct together right after definitions at the beginning of the file? Also, could you submit a PR to https://github.com/awslabs/aws-verification-model-for-libcrypto with these changes on the OpenSLL model?

File: .cbmc-batch/jobs/aws_cryptosdk_md_abort/Makefile [hunk 7]
```diff
@@ -17,7 +17,10 @@ sinclude ../Makefile.local
 #otherwise, use the default values
 include ../Makefile.local_default
 
-UNWINDSET += 
+# We abstract these functions as manual inspection shows there is no pkey associated with the MD_ctx.
+
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_md_abort/Makefile:21]: Remove extra line.

File: .cbmc-batch/jobs/aws_cryptosdk_md_abort/README.md [hunk 8]
```diff
@@ -0,0 +1,3 @@
+# Memory safety proof for aws_cryptosdk_md_abort
+
+This proof harness attains 100% code coverage.
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_md_abort/README.md:3]: If the proof has 100% coverage, you can remove the README file. There is nothing to justify on coverage.

File: .cbmc-batch/jobs/aws_cryptosdk_md_finish/Makefile [hunk 9]
```diff
@@ -17,7 +17,10 @@ sinclude ../Makefile.local
 #otherwise, use the default values
 include ../Makefile.local_default
 
-UNWINDSET += 
+# We abstract these functions as manual inspection shows there is no pkey associated with the MD_ctx.
+
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_md_finish/Makefile:21]: Remove extra line.

File: .cbmc-batch/jobs/aws_cryptosdk_md_init/Makefile [hunk 10]
```diff
@@ -17,7 +17,10 @@ sinclude ../Makefile.local
 #otherwise, use the default values
 include ../Makefile.local_default
 
-UNWINDSET += 
+# We abstract these functions as manual inspection shows there is no pkey associated with the MD_ctx.
+
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_md_init/Makefile:21]: Remove extra line.

File: .cbmc-batch/jobs/aws_cryptosdk_md_init/README.md [hunk 11]
```diff
@@ -0,0 +1,3 @@
+# Memory safety proof for aws_cryptosdk_md_init
+
+This proof harness attains 100% code coverage.
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_md_init/README.md:3]: See previous comment. You can remove this README file.

File: .cbmc-batch/jobs/aws_cryptosdk_md_update/Makefile [hunk 12]
```diff
@@ -28,14 +29,19 @@ ENTRY = aws_cryptosdk_md_update_harness
...
+DEPENDENCIES += $(SRCDIR)/c-common-helper-stubs/error.goto
 DEPENDENCIES += $(SRCDIR)/c-enc-sdk-src/cipher_openssl.goto
 DEPENDENCIES += $(SRCDIR)/helper-src/cbmc_invariants.goto
 DEPENDENCIES += $(SRCDIR)/helper-src/make_common_data_structures.goto
 DEPENDENCIES += $(SRCDIR)/helper-src/openssl/bn_override.goto
 DEPENDENCIES += $(SRCDIR)/helper-src/openssl/ec_override.goto
 DEPENDENCIES += $(SRCDIR)/helper-src/openssl/evp_override.goto
 
+REMOVE_FUNCTION_BODIES += --remove-function-body EVP_MD_CTX_free
+REMOVE_FUNCTION_BODIES += --remove-function-body evp_md_ctx_is_valid
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_md_update/Makefile:41]: Why do you remove the function body for these functions? Add a comment.

File: .cbmc-batch/jobs/aws_cryptosdk_md_update/Makefile [hunk 13]
```diff
@@ -17,7 +17,8 @@ sinclude ../Makefile.local
 #otherwise, use the default values
 include ../Makefile.local_default
 
-UNWINDSET += 
+ABSTRACTIONS += $(SRCDIR)/helper-stubs/EVP_MD_CTX_free_no_pkey_stub.c
+ABSTRACTIONS += $(SRCDIR)/helper-stubs/evp_md_ctx_is_valid_no_pkey_stub.c
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_md_update/Makefile:21]: Why do you abstract these functions? Add a comment.

File: .cbmc-batch/jobs/aws_cryptosdk_md_update/README.md [hunk 14]
```diff
@@ -0,0 +1,3 @@
+# Memory safety proof for aws_cryptosdk_md_update
+
+This proof harness attains 100% code coverage.
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_md_update/README.md:3]: See previous comments. You can remove these README file.

File: .cbmc-batch/stubs/EVP_MD_CTX_free_no_pkey_stub.c [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+ * limitations under the License.
+ */
+
+#include <ec_utils.h>
+#include <make_common_data_structures.h>
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+#include <openssl/kdf.h>
+#include <openssl/rsa.h>
+#include <proof_helpers/nondet.h>
```
Comment (feliperodri) [inline .cbmc-batch/stubs/EVP_MD_CTX_free_no_pkey_stub.c:22]: Why do we need al of these includes here? I think it's enough only the one with the definition for `EVP_MD_CTX `, right?

File: .cbmc-batch/stubs/EVP_MD_CTX_free_no_pkey_stub.c [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+/*
+ * Description: Cleans up digest context ctx and frees up the space allocated to it.
+ * Use this stub when we are *certain* there is no pkey associated with the digest context.
+ */
```
Comment (feliperodri) [inline .cbmc-batch/stubs/EVP_MD_CTX_free_no_pkey_stub.c:27]:
```c
/*
 * Frees up the space allocated to ctx context.
 * Use this stub when we are *certain* there is no pkey associated with the digest context.
 */
```

File: .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+ * limitations under the License.
+ */
+
+#include <ec_utils.h>
+#include <make_common_data_structures.h>
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+#include <openssl/kdf.h>
+#include <openssl/rsa.h>
+#include <proof_helpers/nondet.h>
```
Comment (feliperodri) [inline .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c:22]: Again, we don't need all of these includes, right? Could you remove the unnecessary ones?

File: .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+/* Helper function for CBMC proofs: checks if EVP_MD_CTX is valid.
+ * Use this stub when we are *certain* there is no pkey associated with the digest context.
+ */
```
Comment (feliperodri) [inline .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c:26]:
```c
/*
 * Checks whether EVP_MD_CTX is a valid object.
 * Use this stub when we are certain there is no pkey
 * associated with the digest context.
 */
```

File: .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+bool evp_md_ctx_is_valid(EVP_MD_CTX *ctx) {
+    return ctx && ctx->is_initialized && ctx->digest_size <= EVP_MAX_MD_SIZE && ctx->pkey == NULL;
```
Comment (feliperodri) [inline .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c:28]: If we want to be *certain* to only use this stub when there is no `pkey` associated with the digest context, shouldn't we assert that here?
```c
assert(ctx->pkey == NULL);
```

Comment (feliperodri) [reply .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c:28]: If that isn't the case and we really want to this be a part of the conjunction, update the documentation.

File: source/cipher_openssl.c [hunk 20]
```diff
@@ -161,10 +161,7 @@ int aws_cryptosdk_md_finish(struct aws_cryptosdk_md_context *md_context, void *o
     int rv            = AWS_OP_SUCCESS;
     unsigned int size = 0;
 
-    // Replace with AWS_FATAL_PRECONDITION once that version is integrated
-    if (!output_buf) {
-        abort();
-    }
+    AWS_FATAL_PRECONDITION(output_buf != NULL);
```
Comment (feliperodri) [inline source/cipher_openssl.c:164]: Nice!

**Review by feliperodri — CHANGES_REQUESTED** [review]
A few more comments. You also need to rebase your branch.

File: .cbmc-batch/include/openssl/evp.h (hunk 4 above)
Comment (tegansb) [reply .cbmc-batch/include/openssl/evp.h:28]: Done

File: .cbmc-batch/include/openssl/evp.h (hunk 5 above)
Comment (tegansb) [reply .cbmc-batch/include/openssl/evp.h:54]: Done

File: .cbmc-batch/include/openssl/evp.h (hunk 6 above)
Comment (tegansb) [reply .cbmc-batch/include/openssl/evp.h:59]: Done! And will submit a PR just as soon as I've made the other changes.

File: .cbmc-batch/jobs/aws_cryptosdk_md_abort/Makefile (hunk 7 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_md_abort/Makefile:21]: Done

File: .cbmc-batch/jobs/aws_cryptosdk_md_init/Makefile (hunk 10 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_md_init/Makefile:21]: Done!

File: .cbmc-batch/stubs/EVP_MD_CTX_free_no_pkey_stub.c (hunk 15 above)
Comment (tegansb) [reply .cbmc-batch/stubs/EVP_MD_CTX_free_no_pkey_stub.c:22]: Yeah, I have removed the unnecessary ones!

File: .cbmc-batch/stubs/EVP_MD_CTX_free_no_pkey_stub.c (hunk 16 above)
Comment (tegansb) [reply .cbmc-batch/stubs/EVP_MD_CTX_free_no_pkey_stub.c:27]: Fixed

File: .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c (hunk 17 above)
Comment (tegansb) [reply .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c:22]: Done!

File: .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c (hunk 18 above)
Comment (tegansb) [reply .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c:26]: Done!

File: source/cipher_openssl.c (hunk 20 above)
Comment (tegansb) [reply source/cipher_openssl.c:164]: Thanks!

File: .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c (hunk 19 above)
Comment (tegansb) [reply .cbmc-batch/stubs/evp_md_ctx_is_valid_no_pkey_stub.c:28]: Yes, I think the assert is better and ensures that we really only use this stub when there is no pkey. I have updated this!

File: .cbmc-batch/jobs/aws_cryptosdk_md_abort/README.md (hunk 8 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_md_abort/README.md:3]: Okay, yeah I thought it might be good for documentation but this info is on the spreadsheet so the README adds nothing extra

File: .cbmc-batch/jobs/aws_cryptosdk_md_update/Makefile (hunk 12 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_md_update/Makefile:41]: They shouldn't have been removed. This is now fixed!

File: .cbmc-batch/jobs/aws_cryptosdk_md_update/README.md (hunk 14 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_md_update/README.md:3]: Done!

File: .cbmc-batch/jobs/aws_cryptosdk_md_finish/Makefile (hunk 9 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_md_finish/Makefile:21]: Done!

File: .cbmc-batch/jobs/aws_cryptosdk_md_init/Makefile (hunk 10 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_md_init/Makefile:21]: Done!

File: .cbmc-batch/jobs/aws_cryptosdk_md_init/README.md (hunk 11 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_md_init/README.md]: Done!

**Review by feliperodri — APPROVED** [review]
LGTM.

**Review by danielsn — APPROVED** [review]

File: .cbmc-batch/jobs/aws_cryptosdk_md_abort/Makefile (hunk 1 above)
Comment (agray256) [reply .cbmc-batch/jobs/aws_cryptosdk_md_abort/Makefile]: Verified that the issue raised by tawdry-audrey is resolved. Please update the branch and an admin can override the change request blocking merge.
